### PR TITLE
SI-9560 Remove dependency on parser-combinators/json

### PR DESCRIPTION
**Pull request by soc**
This drops the dependency by integrating the bare minimum functionality
to keep things working.

### CODE REVIEW

**Comment by soc** [comment]
@SethTisue Could you have a look regarding the build/ant/pom details?

File: src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+case class JSONObject (obj: Map[String,Any]) {
+  override def toString = s"{${obj.map({ case (k,v) => format(k) + " : " + format(v) }).mkString(", ")}}"
+}
+
+/**
+ *  Represents a JSON Array (vector).
+ *  @author Derek Chen-Becker <"java"+@+"chen-becker"+"."+"org">
+ */
+case class JSONArray (vector : Vector[Any]) {
+  override def toString = s"[${vector.map(format).mkString(", ")}]"
```
Comment (Ichoran) [inline src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala]: Can we drop the gratuitous string interpolation for `vector.map(format).mkString("[", ", ", "]")`?

File: src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  }
+}
+
+/**
+ *  Represents a JSON Object (map).
+ *
+ *  @author Derek Chen-Becker <"java"+@+"chen-becker"+"."+"org">
+ */
+case class JSONObject (obj: Map[String,Any]) {
+  override def toString = s"{${obj.map({ case (k,v) => format(k) + " : " + format(v) }).mkString(", ")}}"
```
Comment (Ichoran) [inline src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala]: String interpolation isn't helping here.  Just use the 3-arg mkString.

File: src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  /**
+   * The default formatter used by the library. You can
+   * provide your own with the toString calls on
+   * JSONObject and JSONArray instances.
+   */
+  def format(x: Any): String = x match {
+    case s:  String     => s"""\"${quoteString(s)}\""""
+    case jo: JSONObject => jo.toString
+    case ja: JSONArray  => ja.toString
+    case other          => other.toString
```
Comment (Ichoran) [inline src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala]: This is designed to dump all sorts of stuff unquoted into what is supposedly "JSON".  A strict warning is called for at the very least.

File: src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala [hunk 4]
```diff
@@ -0,0 +1,71 @@
+package scala.tools.nsc.doc.html.page
+
+import JSONFormat.format
+
+object JSONFormat {
+
+  /**
+   * The default formatter used by the library. You can
+   * provide your own with the toString calls on
```
Comment (Ichoran) [inline src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala]: But the `toString` calls don't take a formatter?

File: src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala (hunk 2 above)
Comment (som-snytt) [inline src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala]: `obj.map { case (k,v) => s"${format(k)} : ${format(v)}" }.mkString("{", ", ", "}")` to omit extra parens and use `s` inside.

File: src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala (hunk 3 above)
Comment (som-snytt) [inline src/scaladoc/scala/tools/nsc/doc/html/page/JSON.scala]: Triple quotes means backslash is not needed. `s""""${f(s)}""""`.

**Comment by soc** [comment]
@Ichoran Thanks for the feedback!

**Comment by som-snytt** [comment]
My highly nuanced notes were just before the push, BTW. The liberating one is about `"""`.

**Comment by SethTisue** [comment]
@szeiger, you're neck-deep in ant and sbt lately, can you review that aspect?

@soc this looks very promising. did you diff the generated scaladoc for the standard library, before vs. after?

**Comment by soc** [comment]
@SethTisue I diffed a smaller example of my own with all kinds of special characters, would be surprised if there were any differences considering the code is pretty much the same as before.

**Comment by SethTisue** [comment]
I verified that `library/index.js` and `library/lib/index.js` are identical before and after the change. So this LGTM, but I leave the final merge to @szeiger. (I imagine Stefan might want to wait until we're off of Ant altogether, so marking this "on hold" for now.)

**Comment by soc** [comment]
@SethTisue @szeiger What's the current status of the ant -> sbt migration?

**Comment by szeiger** [comment]
Now that https://github.com/scala/scala/pull/4941 has been merged, I think an updated version on top of 2.12.x could be safely merged. The basics of the sbt build should be stable enough not to complicate merges too much in the future. Some tests and integration scripts still need to be ported to sbt (@SethTisue is working on this at the moment), so you'd still need to update both the sbt and the ant build.

**Comment by SethTisue** [comment]
> What's the current status of the ant -> sbt migration

I'm actively working on it. No fundamental obstacles remain afawk, just a parade of details to sort out. The umbrella ticket to watch is https://github.com/scala/scala-dev/issues/45.

**Comment by soc** [comment]
Sorry for the ping ... is there an estimated timeframe when the SBT build will be switched on, or promoted as the main way to build scala/scala?

**Comment by adriaanm** [comment]
We hope to have this done in the next two weeks.

**Comment by lrytz** [comment]
@soc there's no reason to hold this change back, even if we're still using the ant build on CI. could you rebase the patch?

File: src/build/bnd/scala-compiler.bnd [hunk 5]
```diff
@@ -5,7 +5,6 @@ Bundle-Version: ${ver}
 Export-Package: *;version=${ver}
 Import-Package: jline.*;resolution:=optional, \
                 org.apache.tools.ant.*;resolution:=optional, \
-                scala.util.parsing.*;version="${range;[====,====];@PARSER_COMBINATORS_VERSION@}";resolution:=optional, \
```
Comment (szeiger) [inline src/build/bnd/scala-compiler.bnd:8]: There's a similar import spec in `build.sbt` that needs to be changed. This file is only used by ant.

**Comment by szeiger** [comment]
Flushing the M5 queue on Monday. Please update until then if you want this to go into 2.12.0 (but this looks safe to put into a non-.0 release as well)

**Comment by soc** [comment]
@szeiger Updated! Could you have a look?

**Comment by szeiger** [comment]
LGTM

**Comment by SethTisue** [comment]
thanks, Simon! really nice to have this resolved

**Pull request merged**